experimental/overhead_matching/swag/scripts/train.py

```diff
@@ -0,0 +1,195 @@
+    # gradient step.
+    embedding_pool_batch_size: int
+
+    # This is the number of triplets that are processed in a batch.
+    opt_batch_size: int
+
+
+@dataclass
+class TrainConfig:
+    opt_config: OptimizationConfig
+    output_dir: Path
+
+
+@dataclass
+class Triplet:
+    anchor_panorama_idx: int
+    positive_satellite_idx: int
+    negative_satellite_idx: int
+
+
+def create_triplets(
+    panorama_metadata, satellite_metadata, panorama_embeddings, satellite_embeddings
+):
+    # Generate an exhaustive set of triplets where the anchor is a panorama
+    # TODO consider creating triplets where a satellite patch is the anchor
+    out = []
+    for batch_pano_idx in range(panorama_embeddings.shape[0]):
+        anchor_sat_idx = panorama_metadata[batch_pano_idx]["satellite_idx"]
+        # For now, we assume that the panorama and satellite metadata are paired since
+        # the satellite metadata does not currently contain the satellite index
+        for batch_sat_idx in range(panorama_embeddings.shape[0]):
+            neg_sat_idx = panorama_metadata[batch_sat_idx]["satellite_idx"]
+            if anchor_sat_idx == neg_sat_idx:
+                continue
+            out.append(
+                Triplet(
+                    anchor_panorama_idx=batch_pano_idx,
+                    positive_satellite_idx=batch_pano_idx,
+                    negative_satellite_idx=batch_sat_idx,
+                )
+            )
+    return out
+
+
+def compute_loss(anchor_embedding, positive_embeddings, negative_embeddings):
+    return torch.nn.functional.triplet_margin_loss(
+        anchor_embedding, positive_embeddings, negative_embeddings
+    )
+
+
+def train(config: TrainConfig, *, dataset, panorama_model, satellite_model):
+    panorama_model = panorama_model.cuda()
+    satellite_model = satellite_model.cuda()
+
+    opt_config = config.opt_config
+
+    overall_batch_size = (
+        opt_config.embedding_pool_batch_size * opt_config.num_embedding_pool_batches
+    )
+    dataloader = vigor_dataset.get_dataloader(
+        dataset, batch_size=overall_batch_size, num_workers=1
+    )
+
+    opt = torch.optim.Adam(
+        list(panorama_model.parameters()) + list(satellite_model.parameters())
+    )
+
+    for epoch_idx in tqdm.tqdm(range(config.opt_config.num_epochs),  desc="Epoch"):
+        for batch_idx, batch in enumerate(dataloader):
+            panorama_embeddings = []
+            satellite_embeddings = []
+            for i in range(opt_config.num_embedding_pool_batches):
+                start_idx = i * opt_config.embedding_pool_batch_size
+                end_idx = start_idx + opt_config.embedding_pool_batch_size
+                with torch.no_grad():
+                    panos = batch.panorama[start_idx:end_idx].cuda()
+                    satellite_patches = batch.satellite[start_idx:end_idx].cuda()
+                    panorama_embeddings.append(panorama_model(panos).cpu())
```

Comment:
why do we compute the embeddings here?

Reply:
assuming in the future we might want to try picking triplets based on embedding similarity?